File: haskell/p-26.hs


combinations :: (Eq a, Num a) => a -> [t] -> [[t]]
combinations k [] = []
combinations 1 (x:xs) = [[x]] ++ (combinations 1 xs)
combinations k (x:xs) = (map (\ l -> x:l) (combinations (k - 1) xs)) ++
                        (combinations k xs)

File: erlang/p_26.erl
-module(p_26).
-export([combinations/2]).

combinations(_, []) ->
    [];
combinations(1, Xs) ->
    [[X] || X <- Xs];
combinations(M, [X|Xs]) ->
    X1 = [[X|Ys] || Ys <- combinations(M-1, Xs)],
    X2 = [Ys || Ys <- combinations(M, Xs)],
    X1 ++ X2.
